src/lfortran/semantics/ast_common_visitor.h

```
@@ -1340,7 +1340,24 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
             if (m_dim[i].m_start) {
                 this->visit_expr(*m_dim[i].m_start);
                 dim.m_start = ASRUtils::EXPR(tmp);
-            } else {
+                ASR::expr_t* start = dim.m_start;
+                if(ASR::is_a<ASR::Var_t>(*start)) {
+
+                    ASR::Var_t* start_var = ASR::down_cast<ASR::Var_t>(start);
+                    ASR::symbol_t* start_sym = start_var->m_v;
+                    SymbolTable* symbol_scope = ASRUtils::symbol_parent_symtab(start_sym);
+                    if(ASR::is_a<ASR::Variable_t>(*start_sym)) {
+                        ASR::Variable_t* start_variable = ASR::down_cast<ASR::Variable_t>(start_sym);
+
+                        if(start_variable->m_storage != ASR::storage_typeType::Parameter) {
+
+                            if(!in_Subroutine && (symbol_scope->counter == current_scope->counter) && !ASR::is_a<ASR::IntegerConstant_t>(*start)) {
+                                throw SemanticError("Explicit shaped array with nonconstant bounds at ",m_dim[i].m_start->base.loc);
+                            }
+                        }
+                    }
+                }
```

We need to throw a semantic error for the MRE in https://github.com/lfortran/lfortran/issues/5173#issuecomment-2439503478 also. Please add an `else` clause here which checks for `IntegerConstant`.